cmd/kubeadm/app/cmd/upgrade/node.go

```diff
@@ -19,221 +19,253 @@ package upgrade
+	// and keep it here for one year or three releases - the longer of the two - starting from v1.15 included
+	cmd.AddCommand(NewCmdUpgradeNodeConfig())
 
-	options.AddKubeConfigFlag(cmd.Flags(), &flags.kubeConfigPath)
-	cmd.Flags().BoolVar(&flags.dryRun, options.DryRun, flags.dryRun, "Do not change any state, just output the actions that would be performed.")
-	cmd.Flags().BoolVar(&flags.etcdUpgrade, "etcd-upgrade", flags.etcdUpgrade, "Perform the upgrade of etcd.")
-	cmd.Flags().BoolVar(&flags.renewCerts, "certificate-renewal", flags.renewCerts, "Perform the renewal of certificates used by component changed during upgrades.")
 	return cmd
 }
 
-// RunUpgradeNodeConfig is executed when `kubeadm upgrade node config` runs.
-func RunUpgradeNodeConfig(flags *nodeUpgradeFlags) error {
-	if len(flags.kubeletVersionStr) == 0 {
-		return errors.New("the --kubelet-version argument is required")
+// newNodeOptions returns a struct ready for being used for creating cmd kubeadm upgrade node flags.
+func newNodeOptions() *nodeOptions {
+	return &nodeOptions{
+		kubeConfigPath: constants.GetKubeletKubeConfigPath(),
+		dryRun:         false,
 	}
+}
+
+func addUpgradeNodeFlags(flagSet *flag.FlagSet, nodeOptions *nodeOptions) {
+	options.AddKubeConfigFlag(flagSet, &nodeOptions.kubeConfigPath)
+	flagSet.BoolVar(&nodeOptions.dryRun, options.DryRun, nodeOptions.dryRun, "Do not change any state, just output the actions that would be performed.")
+	flagSet.StringVar(&nodeOptions.kubeletVersion, options.KubeletVersion, nodeOptions.kubeletVersion, "The *desired* version for the kubelet config after the upgrade. If not specified, the KubernetesVersion from the kubeadm-config ConfigMap will be used")
+}
 
-	// Set up the kubelet directory to use. If dry-running, use a fake directory
-	kubeletDir, err := upgrade.GetKubeletDir(flags.dryRun)
+// newNodeData returns a new nodeData struct to be used for the execution of the kubeadm upgrade node workflow.
+// This func takes care of validating nodeOptions passed to the command, and then it converts
+// options into the internal InitConfiguration type that is used as input all the phases in the kubeadm upgrade node workflow
+func newNodeData(cmd *cobra.Command, args []string, options *nodeOptions) (*nodeData, error) {
+	client, err := getClient(options.kubeConfigPath, options.dryRun)
 	if err != nil {
-		return err
+		return nil, errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", options.kubeConfigPath)
 	}
 
-	client, err := getClient(flags.kubeConfigPath, flags.dryRun)
-	if err != nil {
-		return errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", flags.kubeConfigPath)
+	// isControlPlane checks if a node is a control-plane node by looking up
+	// the kube-apiserver manifest file
+	isControlPlaneNode := true
+	filepath := kubeadmconstants.GetStaticPodFilepath(kubeadmconstants.KubeAPIServer, kubeadmconstants.GetStaticPodDirectory())
```

Comment:
This exact check is present somewhere in reset (I think) as a separate func. Probably extracting it as an utility func somewhere is a good idea.

Reply:
I will open an issue to track this